### PR TITLE
TypeORM mygration

### PR DESCRIPTION
**Pull request by RomainGuarinoni**
# Whats new ?

## Delete everything related to sql

- the entire `api/sql` folderr has been deleted
- all the commands to create sql tables

If you want to delete the tables created by the sql on your db , run the following commands in your docker container : 

```sql
DROP TABLE IF EXISTS musicians_instruments;
DROP TABLE IF EXISTS events_admin;
DROP TABLE IF EXISTS musicians_genres;
DROP TABLE IF EXISTS events_genres;
DROP TABLE IF EXISTS events_groups;
DROP TABLE IF EXISTS groups_genres;
DROP TABLE IF EXISTS groups_musicians;
DROP TABLE IF EXISTS tokens;
DROP TABLE IF EXISTS musicians;
DROP TABLE IF EXISTS genres;
DROP TABLE IF EXISTS instruments;
DROP TABLE IF EXISTS groups;
DROP TABLE IF EXISTS events;


-- Drop all the type
DROP TYPE IF EXISTS locations;
DROP TYPE IF EXISTS promotions;
DROP TYPE IF EXISTS membership_status;
DROP TYPE IF EXISTS group_roles;
```

To open your docker container, just run the following command : 

```sh
docker exec -i findAMusician_postgres psql postgres://user:pass@postgres:5432/db
```

## Update in generating API types

The function  `api/command/generateType` has been updated so that the `format:'date-time'` in openAPI defintion result in a `Date` type in the `api/types/schema.ts`.

Moreover, openAPI json file is not created anymore ! 

## TypeORM

Every routes now used typeORM to query the db. 

### Add default value in your db

To insert default value in your db, head to your `index.ts` file, and uncomment this line : 

```ts
// await reset();
```
>Sometime, the reset fails, just re-save your file to retrigger nodemon and the reset will work

### Entity

All the entities are located in the `api/entity` folder

### Type checking

Now that we are using typeORM, every query in the db is typed, meaning that each `.find()` 

for example : 

```ts
const musician = await getRepository(Musician).findOne({
        where: { id: req.params.musicianId },
        relations: ['instruments', 'genres'],
      });
```

musician is typed `Musician`.

the fact that the db queries are typed is super powerful because typescript compares the types generated from openAPI and those returned by the db and can bring up errors that we couldn't see before! In conclusion, the backend is much more strongly typed!

Example : 

```ts
router.get(
  '/:musicianId',
  async (
    req: Request<
      getPathParams<GetMusicianById>,
      getResponsesBody<GetMusicianById>,
      getRequestBody<GetMusicianById>,
      {}
    >,
    res: core.Response<
      getResponsesBody<GetMusicianById>,
      {},
      getHTTPCode<GetMusicianById>
    >,
  ) => {
    try {
      const musician = await getRepository(Musician).findOne({
        where: { id: req.params.musicianId },
        relations: ['instruments', 'genres'],
      });

      if (!musician) {
        return res.status(404).json({ msg: 'E_UNFOUND_USER' });
      }

      res.status(200).json(musician);
    } catch (err) {
      console.log(err);
      return res
        .status(500)
        .json({ msg: 'E_SQL_ERROR', stack: JSON.stringify(err) });
    }
  },
);
```

for example, typescript compare the type of `musician` to `getResponsesBody<GetMusicianById>`  ( the type generated by openAPI) and if there is an error, fir example a field is named differently between openAPI and typeorm, typescript will create an error ! 

## Open API changes

There is a lot of openAPI definition changes in this PR because sometimes, openAPI definition were written in the wrong way and typescripts was creating errors! But the behavior of every routes is not changed

### CODE REVIEW

File: api/entity/Genre.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    onDelete: 'CASCADE',
+  })
+  musicians: Musician[];
+
+  @ManyToMany(() => Band, (band) => band.genres, { onDelete: 'CASCADE' })
+  groups: Band[];
+
+  @ManyToMany(() => Event, (event) => event.genres, { onDelete: 'CASCADE' })
+  events: Event[];
+}
```
Comment (alexandrelam) [inline api/entity/Genre.ts:24]: I don't think this belongs here. If we want to get infos on which bands have specific group or events we should do nested routes. We'll also have problems down the line with cascade delete. WDYT?

**Review by alexandrelam — COMMENTED** [review]
Shouldn't we rename [api/entity/Band.ts](https://github.com/Find-A-Musician/findAMusicianBackend/pull/78/files#diff-27b19990bb5ee630298d3cff1a4263a595429666a7595afa4f9dae6b55331949) to `groups` for consistency sake?

File: api/entity/Instrument.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  @ManyToMany(() => Musician, (musician) => musician.instruments, {
+    onDelete: 'CASCADE',
+  })
+  musicians: Musician[];
+
+  @ManyToMany(() => MusicianBand, (musicianBand) => musicianBand.instruments, {
+    onDelete: 'CASCADE',
+  })
+  groups: MusicianBand;
+}
```
Comment (alexandrelam) [inline api/entity/Instrument.ts:22]: Same reasoning as `Genre.ts`. I don't think this belongs here. We should just remove it and create nested routes.

File: api/routes/events/events.ts [hunk 3]
```diff
@@ -71,34 +93,31 @@ router.post(
...
+      const admin = new Musician();
+      admin.id = req.userId;
+
+      const newEvent = new Event();
+      newEvent.name = name;
+      newEvent.description = description;
+      newEvent.startDate = startDate;
+      newEvent.endDate = endDate;
+      newEvent.adress = adress;
+      newEvent.admins = [admin];
```
Comment (alexandrelam) [inline api/routes/events/events.ts:108]:
```suggestion
      const newEvent = new Event({name, description, startDate, endDate, adress, admin});
```
Can't you just do something like that?

**Review by alexandrelam — CHANGES_REQUESTED** [review]

File: api/entity/Genre.ts (hunk 1 above)
Comment (RomainGuarinoni) [reply api/entity/Genre.ts:24]: You're right ! I'm going to change that

Comment (RomainGuarinoni) [reply api/entity/Genre.ts]: Done

File: api/routes/events/events.ts (hunk 3 above)
Comment (RomainGuarinoni) [reply api/routes/events/events.ts:108]: We can if we create constructor for each entity, i will do that

Comment (RomainGuarinoni) [reply api/routes/events/events.ts]: Done without create constructors but instead just calling the `getRepository(T).create<T>()`

**Review by alexandrelam — APPROVED** [review]
Too much code I trust you. It should mainly be good though. If there are bugs we'll fix them in other prs.

**Review by AlexisAoun — APPROVED** [review]
Less SQL can only be good. sgtm

**Comment by AlexisAoun** [comment]
Should we push the unti test changes on this PR as well ? As it is a part of the migration

**Comment by AlexisAoun** [comment]
> > Should we push the unti test changes on this PR as well ? As it is a part of the migration
> 
> We should merge now. Front need the new back with the new tables. Fix the tests later in a new pr please.

Understandable, lets merge then

**Pull request merged**